File: notification_service/src/notification/application/services/notification_service.py
import logging
from typing import Any
from uuid import UUID

from fern_labour_core.unit_of_work import UnitOfWork
from fern_labour_notifications_shared.enums import NotificationTemplate

from src.core.application.domain_event_publisher import DomainEventPublisher
from src.core.domain.domain_event.repository import DomainEventRepository
from src.notification.application.dtos.notification import NotificationDTO
from src.notification.application.services.notification_generation_service import (
    NotificationGenerationService,
)
from src.notification.application.services.notification_router import NotificationRouter
from src.notification.domain.entity import Notification
from src.notification.domain.enums import (
    NotificationChannel,
    NotificationStatus,
)
from src.notification.domain.exceptions import (
    InvalidNotificationChannel,
    InvalidNotificationId,
    InvalidNotificationStatus,
    InvalidNotificationTemplate,
    NotificationNotFoundById,
)
from src.notification.domain.repository import NotificationRepository
from src.notification.domain.services.can_resend_notification_service import (
    CanResendNotificationService,
)
from src.notification.domain.value_objects.notification_id import NotificationId

log = logging.getLogger(__name__)


class NotificationService:
    def __init__(
        self,
        notification_router: NotificationRouter,
        notification_generation_service: NotificationGenerationService,
        notification_repository: NotificationRepository,
        domain_event_repository: DomainEventRepository,
        domain_event_publisher: DomainEventPublisher,
        unit_of_work: UnitOfWork,
    ):
        self._notification_router = notification_router
        self._notification_generation_service = notification_generation_service
        self._notification_repository = notification_repository
        self._domain_event_repository = domain_event_repository
        self._domain_event_publisher = domain_event_publisher
        self._unit_of_work = unit_of_work

    async def _get_notification(self, notification_id: str) -> Notification:
        try:
            domain_notification_id = NotificationId(UUID(notification_id))
        except ValueError:
            raise InvalidNotificationId(notification_id=notification_id)

        notification = await self._notification_repository.get_by_id(
            notification_id=domain_notification_id
        )
        if not notification:
            raise NotificationNotFoundById(notification_id=notification_id)

        return notification

    async def _send_notification(self, notification: Notification) -> None:
        notification_content = await self._notification_generation_service.generate_content(
            notification=notification
        )

        notification_dto = NotificationDTO.from_domain(notification=notification)
        notification_dto.add_notification_content(content=notification_content)

        result = await self._notification_router.route_notification(notification_dto)

        if result.external_id:
            notification.external_id = result.external_id

        notification.update_status(result.status)

        async with self._unit_of_work:
            await self._notification_repository.save(notification=notification)
            await self._domain_event_repository.save_many(notification.clear_domain_events())

        self._domain_event_publisher.publish_batch_in_background()

    async def create_notification(
        self,
        channel: str,
        destination: str,
        template: str,
        data: dict[str, Any],
        status: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> NotificationDTO:
        try:
            notification_channel = NotificationChannel(channel)
        except ValueError:
            raise InvalidNotificationChannel(notification_channel=channel)

        try:
            notification_status = NotificationStatus(status) if status else None
        except ValueError:
            raise InvalidNotificationStatus(notification_status=status)

        try:
            notification_template = NotificationTemplate(template)
        except ValueError:
            raise InvalidNotificationTemplate(template=template)

        notification = Notification.create(
            channel=notification_channel,
            destination=destination,
            template=notification_template,
            data=data,
            metadata=metadata,
            status=notification_status,
        )

        async with self._unit_of_work:
            await self._notification_repository.save(notification)
            await self._domain_event_repository.save_many(notification.clear_domain_events())

        self._domain_event_publisher.publish_batch_in_background()

        return NotificationDTO.from_domain(notification)

    async def send(self, notification_id: str) -> None:
        notification = await self._get_notification(notification_id=notification_id)
        await self._send_notification(notification=notification)

    async def resend(self, notification_id: str) -> None:
        notification = await self._get_notification(notification_id=notification_id)
        CanResendNotificationService().can_resend_notification(notification=notification)
        await self._send_notification(notification=notification)

File: rust_notification_service/src/infrastructure/services/notification_service.rs
use std::{collections::HashMap, str::FromStr, sync::Arc};

use async_trait::async_trait;
use uuid::Uuid;

use crate::{
    application::{
        dtos::notification::NotificationDTO,
        exceptions::AppError,
        services::{
            notification_generation_service::NotificationGenerationServiceTrait,
            notification_service::NotificationServiceTrait,
        },
    },
    domain::{
        entity::Notification,
        enums::{NotificationChannel, NotificationStatus, NotificationTemplate},
        exceptions::InvalidNotificationStatus,
        repository::NotificationRepository as NotificationRepositoryInterface,
    },
};

pub struct NotificationService {
    pub repo: Arc<dyn NotificationRepositoryInterface + Send + Sync>,
    pub notification_generation_service: Arc<dyn NotificationGenerationServiceTrait>,
}

impl NotificationService {
    pub fn create(
        notification_repo: Arc<dyn NotificationRepositoryInterface>,
        notification_generation_service: Arc<dyn NotificationGenerationServiceTrait>,
    ) -> Arc<dyn NotificationServiceTrait> {
        Arc::new(Self {
            repo: notification_repo,
            notification_generation_service,
        })
    }

    async fn _get_notification(
        &self,
        notification_id: &str,
    ) -> Result<Option<Notification>, AppError> {
        let notification_id = match Uuid::try_parse(&notification_id) {
            Ok(id) => id,
            Err(_) => return Err(AppError::InvalidNotificationId(notification_id.to_string())),
        };
        match self.repo.get_by_id(notification_id).await {
            Ok(notification) => Ok(notification),
            Err(err) => Err(AppError::DatabaseError(err)),
        }
    }

    async fn _get_notification_by_external_id(
        &self,
        external_id: &str,
    ) -> Result<Option<Notification>, AppError> {
        match self.repo.get_by_external_id(external_id).await {
            Ok(notification) => Ok(notification),
            Err(err) => Err(AppError::DatabaseError(err)),
        }
    }
}

#[async_trait]
impl NotificationServiceTrait for NotificationService {
    async fn create_notification(
        &self,
        channel: String,
        destination: String,
        template: String,
        data: HashMap<String, String>,
        metadata: Option<HashMap<String, String>>,
        status: Option<String>,
    ) -> Result<NotificationDTO, AppError> {
        let channel = NotificationChannel::from_str(&channel);
        if let Err(error) = channel {
            return Err(AppError::ValidationError(error.to_string()));
        }
        let channel = channel.unwrap();

        let notification_status: Result<Option<NotificationStatus>, InvalidNotificationStatus> =
            Ok(None);
        if status.is_some() {
            let notification_status = NotificationStatus::from_str(&status.unwrap());
            if let Err(error) = notification_status {
                return Err(AppError::ValidationError(error.to_string()));
            }
        }
        let notification_status = notification_status.unwrap();

        let notification_template = NotificationTemplate::from_str(&template);
        if let Err(error) = notification_template {
            return Err(AppError::ValidationError(error.to_string()));
        }
        let notification_template = notification_template.unwrap();

        let notification = Notification::create(
            channel,
            destination,
            notification_template,
            data,
            notification_status,
            metadata,
        );

        if let Err(err) = self.repo.save(&notification).await {
            tracing::error!("Error creating notification: {err}");
            return Err(AppError::DatabaseError(err));
        }

        Ok(NotificationDTO::from(notification))
    }

    async fn update_notification(
        &self,
        notification_id: String,
        status: String,
        external_id: Option<String>,
    ) -> Result<NotificationDTO, AppError> {
        let notification_status = NotificationStatus::from_str(&status);
        if let Err(error) = notification_status {
            return Err(AppError::ValidationError(error.to_string()));
        }
        let notification_status = notification_status.unwrap();

        let mut notification = match self._get_notification(&notification_id).await? {
            Some(notification) => notification,
            None => {
                return Err(AppError::NotFound(format!(
                    "Notification not found by id: {}",
                    notification_id
                )));
            }
        };

        if let Some(external_id) = external_id {
            notification.set_external_id(external_id);
        };
        notification.set_status(notification_status);

        if let Err(err) = self.repo.save(&notification).await {
            tracing::error!("Error saving notification: {err}");
            return Err(AppError::DatabaseError(err));
        }

        Ok(NotificationDTO::from(notification))
    }

    async fn status_callback(&self, external_id: String, status: String) -> Result<bool, AppError> {
        let notification_status = NotificationStatus::from_str(&status);
        if let Err(error) = notification_status {
            return Err(AppError::ValidationError(error.to_string()));
        }
        let notification_status = notification_status.unwrap();

        let mut notification = match self._get_notification_by_external_id(&external_id).await? {
            Some(notification) => notification,
            None => {
                return Err(AppError::NotFound(format!(
                    "Notification not found by external id: {}",
                    external_id
                )));
            }
        };

        notification.set_status(notification_status);

        if let Err(err) = self.repo.save(&notification).await {
            tracing::error!("Error saving notification: {err}");
            return Err(AppError::DatabaseError(err));
        }

        Ok(true)
    }
}
